### PR TITLE
[PREVIEW] =clu #23286 filter emitted reachability event by DC

### PR DESCRIPTION
**Pull request by jrudolph**
I don't think that just this can work but I'd like to see what breaks.

### CODE REVIEW

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 1]
```diff
@@ -450,9 +450,12 @@ private[cluster] final class ClusterDomainEventPublisher extends Actor with Acto
...
 
   def publishDiff(oldState: MembershipState, newState: MembershipState, pub: AnyRef ⇒ Unit): Unit = {
+    def inSameDc(reachabilityEvent: ReachabilityEvent): Boolean =
+      reachabilityEvent.member.dataCenter == selfDc
+
     diffMemberEvents(oldState, newState) foreach pub
-    diffUnreachable(oldState, newState) foreach pub
-    diffReachable(oldState, newState) foreach pub
+    diffUnreachable(oldState, newState) filter inSameDc foreach pub
+    diffReachable(oldState, newState) filter inSameDc foreach pub
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: um, that is not enough :)

In the end you should find `Reachability.Record` entries with `observer.dataCenter == selfDc && subject.dataCenter == selfDc`. All other records should be thrown away.

Perhaps something like this in `MembershipState`:
```
dcReachability.remove(members.collect { case m if m.dataCenter != selfDc ⇒ m.uniqueAddress })
```

**Comment by patriknw** [comment]
Here we must also remember to use the right `unreachable` when creating `CurrentClusterState`, see `sendCurrentClusterState`

**Comment by jrudolph** [comment]
I see, will add that as well.

**Comment by jrudolph** [comment]
Updated to to filter observations instead of filtering the events.

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 2]
```diff
@@ -450,6 +447,9 @@ private[cluster] final class ClusterDomainEventPublisher extends Actor with Acto
   }
 
   def publishDiff(oldState: MembershipState, newState: MembershipState, pub: AnyRef ⇒ Unit): Unit = {
+    def inSameDc(reachabilityEvent: ReachabilityEvent): Boolean =
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: not used

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala [hunk 3]
```diff
@@ -88,7 +88,7 @@ abstract class MultiDcSpec
         testConductor.blackhole(first, third, Direction.Both).await
       }
       runOn(first, second, third, fourth) {
-        awaitAssert(clusterView.unreachableMembers should not be empty)
+        awaitAssert(clusterView.reachability.allUnreachable should not be empty)
       }
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala]: To add some more integration style of test for this we can add a `cluster.subscribe` to a TestProbe here and see that we get the expected events. Well here it shouldn't get any `UnreachableMember` (which is also a good test) but it should get `UnreachableDataCenter` when @2m has finished that.

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 4]
```diff
@@ -268,13 +268,12 @@ object ClusterEvent {
...
       val newGossip = newState.latestGossip
-      val oldUnreachableNodes = oldGossip.overview.reachability.allUnreachableOrTerminated
-      (newGossip.overview.reachability.allUnreachableOrTerminated.collect {
+      val oldUnreachableNodes = oldState.dcReachabilityNoOutsideNodes.allUnreachableOrTerminated
+      newState.dcReachabilityNoOutsideNodes.allUnreachableOrTerminated.collect {
         case node if !oldUnreachableNodes.contains(node) && node != newState.selfUniqueAddress ⇒
           UnreachableMember(newGossip.member(node))
-      })(collection.breakOut)
+      }(collection.breakOut)
     }
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: perhaps add some test in `ClusterDomainEventSpec` to verify that events are not emitted for some cross cases (it probably already covers the "within" cases)

**Review by patriknw — COMMENTED** [review]
looking good, perhaps add some more tests

**Comment by johanandren** [comment]
I'll pick this up and get it mergeable.

**Pull request closed without merging**